provider/aws: Allow account ID checks on EC2 instances & w/ federated accounts

radeksimko opened the pull request:
I'm building on top of @ColinHebert 's [PR](https://github.com/hashicorp/terraform/pull/4290) and also [recently merged PR](https://github.com/aws/aws-sdk-go/pull/512) in AWS SDK which made this _fairly clean_ approach possible.

This is closing https://github.com/hashicorp/terraform/issues/3431

Here's an overview of the implemented approaches per provider:

| provider/approaches       	| 1st          	| 2nd           	| 3rd 	|
|---------------------------	|--------------	|---------------	|-----	|
| EC2RoleProvider           	| Metadata API 	| ERR           	|     	|
| EnvProvider               	| iam:GetUser  	| iam:ListRoles 	| ERR 	|
| SharedCredentialsProvider 	| iam:GetUser  	| iam:ListRoles 	| ERR 	|
| StaticCredentialsProvider 	| iam:GetUser  	| iam:ListRoles 	| ERR 	|
| CustomProvider            	| iam:GetUser  	| iam:ListRoles 	| ERR 	|

I would love to add some tests for `getAccountId()` but mocking the IAM service seems quite tricky at the moment.

I have tested all of the providers (launched EC2 instance w/ instance profile + test all other providers each locally).

ColinHebert commented on builtin/providers/aws/auth_helpers.go:
```diff
@@ -0,0 +1,170 @@
+}
+
+func buildMetadataUrl(path string) string {
+	url := os.Getenv("AWS_METADATA_URL")
```
If it's already deprecated, is it worth checking for this variable at all?

ColinHebert commented on builtin/providers/aws/auth_helpers.go:
```diff
@@ -0,0 +1,170 @@
+	if url != "" {
+		log.Printf("[WARN] Using deprecated ENV variable AWS_METADATA_URL, " +
+			"use AWS_METADATA_HOST instead to set the hostname only")
+		return url
```
Will this work? Don't we need to add `/meta-data/iam/info`? `AWS_METADATA_URL` used to be `ts.URL+"/latest"`
If it's deprecated, could failing completely here be an option?

radeksimko replied:
Well, my assumption is that this has only been used in tests and never documented, so I was even thinking of just removing this ENV variable. Erroring out is another option. Unfortunately we don't have any nice way to return warnings from this scope.

radeksimko replied:
> `AWS_METADATA_URL` used to be `ts.URL+"/latest"`

Correct, that is why am I treating it differently and just take the whole URL, instead of building it.
That function is used in two places, have a look into the full diff - we call this API first when deciding whether to use EC2Role provider at all and then I used it here (with different path) to get the role ARN and Account ID from it.

ColinHebert replied:
I'd be all for removing that Env variable, if it isn't documented (plus given the nature of the variable) it's safe to assume it hasn't been used and is safe to remove.

ColinHebert commented on builtin/providers/aws/auth_helpers.go:
```diff
@@ -0,0 +1,170 @@
+		return url
+	}
+
+	host := os.Getenv("AWS_METADATA_HOST")
```
is `AWS_METADATA_HOST` just `169.254.169.254`? Should there be a `http://` before?

radeksimko replied:
There is actually `http://` in the default host. See line `22`.

Maybe the variable name is misleading then as you might not expect `_HOST` to contain the protocol? Any ideas on how to improve this to make it more readable/obvious? `AWS_METADATA_BASE_URI` or something else?

ColinHebert replied:
Ah, gotcha. Yes, `BASE_URI` would make it clearer. Also it should be documented.

radeksimko replied:
Ok, I will change the name & document it.

radeksimko replied:
Ok, I will change the name & document it.

ColinHebert commented on builtin/providers/aws/auth_helpers.go:
```diff
@@ -0,0 +1,170 @@
+	return awsCredentials.NewChainCredentials(providers)
+}
+
+func buildMetadataUrl(path string) string {
```
The more I look at this function, the more I think it should be `getMetadataUrl()` (no parameters), and it should simply return `AWS_METADATA_URL`, `http:// + AWS_METADATA_HOST + /latest` or `DefaultMetadataApiHost + /latest`.

I'm not quite convinced that we need to handle that many different ways to set up the metadata URL.

radeksimko replied:
It's probably not a bad idea. :+1: 

However I'd be statically returning `/latest/meta-data/iam/info` instead of `/latest` and change the function name accordingly.

IAM is only available when Instance Profile is attached, whereas `/latest` is available on all EC2 instances. EC2 instances without instance profiles is practically useless for Terraform (as we won't get any credentials), so we always want `/latest/meta-data/iam/info`.

radeksimko replied:
`ec2metadata.EC2Metadata` is better solution as you mentioned for this context here.

However I will change the current occurence of `/latest` to `/latest/meta-data/iam/info` anyway. we cannot use the `ec2metadata.EC2Metadata` in the context of building ChainProvider as we don't have it available yet at that point.

ColinHebert commented on builtin/providers/aws/auth_helpers.go:
```diff
@@ -0,0 +1,170 @@
+
+func getAccountId(iamconn *iam.IAM, authProviderName string) (string, error) {
+	// If we have creds from instance profile, we can use metadata API
+	if authProviderName == "EC2RoleProvider" {
```
We might want to rely on the constants of the aws-sdk

radeksimko replied:
Good point! I'll change that.

ColinHebert commented on builtin/providers/aws/auth_helpers.go:
```diff
@@ -0,0 +1,170 @@
+	}
+
+	// Then try IAM GetUser
+	log.Println("[DEBUG] Trying to get account ID via iam:GetUser")
```
can we if/elseif on the other provider names? This should remove the need to poke the AWS API if we're not using User credentials

radeksimko replied:
Can you be more specific? I'm not sure I follow. What would be in those if/elseif and what API calls in which specific cases would we save?

ColinHebert replied:
Ah nevermind, I was thinking we can exclude `AssumeRoleProvider` from the `GetUser` path, but it's more cumbersome than helpful

ColinHebert commented on builtin/providers/aws/auth_helpers.go:
```diff
@@ -0,0 +1,170 @@
+	// If we have creds from instance profile, we can use metadata API
+	if authProviderName == "EC2RoleProvider" {
+		log.Println("[DEBUG] Trying to get account ID via AWS Metadata API")
+		url := buildMetadataUrl("/latest/meta-data/iam/info")
```
Could we rely on `ec2metadata.EC2Metadata` for this?

radeksimko replied:
That is a very good idea! That can make the code a bit cleaner. :smiley:

ColinHebert commented on builtin/providers/aws/auth_helpers.go:
```diff
@@ -0,0 +1,170 @@
+
+	// We only look in the EC2 metadata API if we can connect
+	// to the metadata service within a reasonable amount of time
+	metadataURL := buildMetadataUrl("/latest")
```
I think that `ec2metadata.EC2Metadata` might reduce the amount of code required here.

radeksimko replied:
That's not a bad idea, but we'll have to check what's the default timeout. I assume there was a reason to add this "workaround" instead of leaving the EC2Role provider in the chain for auto-detection.

radeksimko commented:
@ColinHebert I think I addressed all of your comments. I have also separated the PR into more commits to make each refactoring phase isolated and more readable. Let me know if you see anything else.

Pinging @catsby as he was behind the recent December refactoring of this. Can you specifically look at eeea8648e4cd9bc493a7ec6381a5022a71f5971b , please? As @ColinHebert suggested, this seems to be much cleaner approach to address that problem with metadata API timeout (which is 5 secs by default).

btw. after this refactoring which makes the code a bit cleaner, there's one concern left, see https://github.com/aws/aws-sdk-go/issues/543

ColinHebert commented:
This is so awesome! Nice work @radeksimko!

radeksimko commented:
This is currently on hold.

I'm waiting for response from AWS in regards to suggestions on best way to verify EC2 metadata API endpoint: https://github.com/aws/aws-sdk-go/issues/543 , https://forums.aws.amazon.com/thread.jspa?threadID=226140

radeksimko commented:
Based on my discussion with AWS folks over in https://github.com/aws/aws-sdk-go/issues/543 about ways to verify EC2 metadata endpoint I modified this PR slightly.

I also submitted https://github.com/aws/aws-sdk-go/pull/598 which should hopefully make the diff in this PR even smaller (LOC that can be in the SDK instead).

I may send another PR to the SDK, just a simple `ec2metadata.AccountId()` helper, but I don't want it to be blocking this PR. I think I have reached a point where it's fairly clean and can be reviewed.

catsby commented:
Sorry for the silence here @radeksimko . It looks like aws/aws-sdk-go#598 was merged and went out in the v1.1.13, do you want to refactor here?

radeksimko commented:
@catsby I will refactor this and then retest on real EC2 instance and w/ real federated credentials.

radeksimko commented:
This is now refactored. I also spent some time creating mock IAM server which allowed me writing couple of tests.

Besides those tests I also tested this manually with real federated credentials and in a real EC2 instance w/ instance profile. Both worked ok.

I also exposed both methods (`GetCredentials` and `GetAccountId`) as public to make it useful for https://github.com/hashicorp/terraform/pull/5270

radeksimko commented:
This is now squashed into 2 commits (vendor + actual changes), as discussed w/ @catsby 

The plan is to review & merge this _after_ the next release (`0.6.15`). This is to give more time for testing "in the wild" to allow capturing any potential edge-cases/bugs that may not have been covered by tests.

catsby commented:
LGTM, unit tests and I ran some acceptance tests locally which passed, I think we're good to merge here thanks!

radeksimko commented:
Just FYI: Although account ID checking by itself does work with EC2 instances and federated credentials now, Terraform still calls `iam:GetUser` as part of credentials validation. This still fails on EC2 & w/ federated credentials.

Feel free to subscribe to https://github.com/hashicorp/terraform/issues/6523 where we track this issue.

radeksimko commented:
Actually, apologies, what I said is not entirely true, it does work for some cases which can be characterised by the following condition:

```go
if awsErr.Code() == "AccessDenied" || awsErr.Code() == "ValidationError" {
```

The linked issue & upcoming PR will hopefully _actually validate_ the credentials at all cases instead of ignoring the check though, so the note is still partially valid.
